File: scripts/mk.py
# -*- coding: utf-8 -*-
"""mk.py new <arq> <titulo> <desc> <lead> <link-antes>   |   mk.py add <arq>"""
import re, glob, sys, os
SITE = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
BLOCO = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'blocos')
os.chdir(SITE)

def topicos_toc(partes):
    out = []
    for sec in re.finditer(r'<section class="parte" id="(s\d+)">(.*?)\n</section>', partes, re.S):
        sid, corpo = sec.group(1), sec.group(2)
        num = re.search(r'PARTE (\d+)</span><h2>(.*?)</h2>', corpo)
        out.append('<a class="sumario__parte" href="#%s">%s · %s</a>' % (sid, num.group(1), num.group(2)))
        out.append('<ol>')
        for t in re.finditer(r'<article class="topico" id="([^"]+)"[^>]*>\s*<div class="topico__cabeca"><h3>(.*?)</h3>', corpo, re.S):
            out.append('<li><a href="#%s">%s</a></li>' % (t.group(1), t.group(2)))
        out.append('</ol>')
    return '\n'.join(out)

def novas_partes():
    fs = sorted(glob.glob(BLOCO + '/p*.html'), key=lambda f: int(re.search(r'p(\d+)', f).group(1)))
    return ''.join(open(f, encoding='utf-8').read() for f in fs)

modo = sys.argv[1]
partes = novas_partes()

if modo == 'new':
    arq, titulo, desc, lead, antes = sys.argv[2:7]
    rotulo = sys.argv[7]
    link = '<a href="%s"><strong>%s</strong><span>%s</span></a>' % (arq, titulo, rotulo)
    modelo = open('movimento.html', encoding='utf-8').read()
    menu = re.search(r'<nav class="navmenu".*?</nav>', modelo, re.S).group(0)
    menu = menu.replace(' class="is-atual"', '')
    menu = re.sub(r'<span class="navmenu__aqui">.*?</span>', '', menu, flags=re.S)
    menu = menu.replace('<a href="%s">' % antes, link + '\n<a href="%s">' % antes)

    for f in glob.glob('*.html'):
        if f in (arq, 'artifact-index.html'): continue
        h = open(f, encoding='utf-8').read()
        if arq in h: continue
        h = re.sub(r'(<a href="%s"(?: class="is-atual")?><strong>)' % antes, link + r'\n\1', h, count=1)
        open(f, 'w', encoding='utf-8').write(h)

    mp = menu.replace('<a href="%s">' % arq, '<a href="%s" class="is-atual">' % arq)
    mp = mp.rstrip()[:-6] + '<span class="navmenu__aqui">%s</span>\n</nav>' % titulo
    np_, nt = partes.count('class="parte"'), partes.count('class="topico"')
    pag = open(BLOCO + '/molde.html', encoding='utf-8').read()
    pag = pag.replace('{{T}}', titulo).replace('{{D}}', desc).replace('{{LEAD}}', lead)
    pag = pag.replace('{{MENU}}', mp).replace('{{SUM}}', '<p class="sumario__titulo">%s</p>\n' % titulo + topicos_toc(partes))
    pag = pag.replace('{{NP}}', str(np_)).replace('{{NT}}', str(nt)).replace('{{PARTES}}', partes)
    open(arq, 'w', encoding='utf-8').write(pag)
    print('%s: %d partes, %d topicos' % (arq, np_, nt))

elif modo == 'add':
    arq = sys.argv[2]
    h = open(arq, encoding='utf-8').read()
    ultimo = max(int(x) for x in re.findall(r'<section class="parte" id="s(\d+)">', h))
    ultimo_num = max(int(x) for x in re.findall(r'PARTE (\d+)</span>', h))
    # renumera o bloco novo
    for i, sec in enumerate(re.findall(r'<section class="parte" id="s(\d+)">', partes)):
        partes = partes.replace('id="s%s">' % sec, 'id="s@%d">' % (ultimo + i + 1), 1)
    for i, n in enumerate(re.findall(r'PARTE (\d+)</span>', partes)):
        partes = partes.replace('PARTE %s</span>' % n, 'PARTE @%d</span>' % (ultimo_num + i + 1), 1)
    partes = partes.replace('id="s@', 'id="s').replace('PARTE @', 'PARTE ')
    h = h.replace('\n</main>', '\n\n' + partes + '\n</main>')
    # sumário
    novo_toc = topicos_toc(partes)
    h = h.replace('</div>\n</nav>\n<main class="leitura">', novo_toc + '\n</div>\n</nav>\n<main class="leitura">')
    np_, nt = h.count('class="parte"'), h.count('class="topico"')
    h = re.sub(r'<span>\d+ partes? · \d+ tópicos</span>', '<span>%d partes · %d tópicos</span>' % (np_, nt), h)
    open(arq, 'w', encoding='utf-8').write(h)
    print('%s: agora %d partes, %d topicos' % (arq, np_, nt))

File: scripts/mk.mjs
/* ============================================================
   mk.mjs new <arq> <titulo> <desc> <lead> <link-antes> <rotulo>
   mk.mjs add <arq>
   Porta de mk.py (Python não está disponível nesta máquina).

   Monta uma trilha nova (ou adiciona partes a uma existente) a
   partir dos blocos em scripts/blocos/p*.html e do molde em
   scripts/blocos/molde.html — nenhum dos dois está neste repositório
   no momento, então "new" e "add" só funcionam se essa pasta existir.
   ============================================================ */
import { readdirSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { lerTexto, escreverTexto } from './_texto.mjs';

const AQUI = dirname(fileURLToPath(import.meta.url));
const SITE = join(AQUI, '..');
const BLOCO = join(AQUI, 'blocos');
process.chdir(SITE);

function topicosToc(partes) {
  const out = [];
  for (const sec of partes.matchAll(/<section class="parte" id="(s\d+)">(.*?)\n<\/section>/gs)) {
    const [, sid, corpo] = sec;
    const num = corpo.match(/PARTE (\d+)<\/span><h2>(.*?)<\/h2>/);
    if (!num) continue;
    out.push(`<a class="sumario__parte" href="#${sid}">${num[1]} · ${num[2]}</a>`);
    out.push('<ol>');
    for (const t of corpo.matchAll(
      /<article class="topico" id="([^"]+)"[^>]*>\s*<div class="topico__cabeca"><h3>(.*?)<\/h3>/gs
    )) {
      out.push(`<li><a href="#${t[1]}">${t[2]}</a></li>`);
    }
    out.push('</ol>');
  }
  return out.join('\n');
}

function novasPartes() {
  let entradas;
  try {
    entradas = readdirSync(BLOCO, { withFileTypes: true });
  } catch {
    return ''; // pasta scripts/blocos/ não existe — mesmo resultado que glob.glob teria
  }
  const arquivos = entradas
    .filter((e) => e.isFile() && /^p\d+\.html$/.test(e.name))
    .map((e) => join(BLOCO, e.name))
    .sort((a, b) => Number(a.match(/p(\d+)/)[1]) - Number(b.match(/p(\d+)/)[1]));
  return arquivos.map((f) => lerTexto(f)).join('');
}

const args = process.argv.slice(2);
const modo = args[0];
const partes0 = novasPartes();

if (modo === 'new') {
  const [arq, titulo, desc, lead, antes] = args.slice(1, 6);
  const rotulo = args[6];
  const link = `<a href="${arq}"><strong>${titulo}</strong><span>${rotulo}</span></a>`;

  const modelo = lerTexto('movimento.html');
  let menu = modelo.match(/<nav class="navmenu".*?<\/nav>/s)[0];
  menu = menu.replaceAll(' class="is-atual"', '');
  menu = menu.replace(/<span class="navmenu__aqui">.*?<\/span>/s, '');
  menu = menu.replaceAll(`<a href="${antes}">`, `${link}\n<a href="${antes}">`);

  const htmlFiles = readdirSync('.', { withFileTypes: true })
    .filter((e) => e.isFile() && e.name.endsWith('.html'))
    .map((e) => e.name);
  for (const f of htmlFiles) {
    if (f === arq || f === 'artifact-index.html') continue;
    let h = lerTexto(f);
    if (h.includes(arq)) continue;
    const re = new RegExp(`(<a href="${antes}"(?: class="is-atual")?><strong>)`);
    h = h.replace(re, (m, g1) => `${link}\n${g1}`);
    escreverTexto(f, h);
  }

  let mp = menu.replaceAll(`<a href="${arq}">`, `<a href="${arq}" class="is-atual">`);
  mp = mp.trimEnd().slice(0, -6) + `<span class="navmenu__aqui">${titulo}</span>\n</nav>`;
  const np_ = (partes0.match(/class="parte"/g) || []).length;
  const nt = (partes0.match(/class="topico"/g) || []).length;

  let pag = lerTexto(join(BLOCO, 'molde.html'));
  pag = pag.replaceAll('{{T}}', titulo).replaceAll('{{D}}', desc).replaceAll('{{LEAD}}', lead);
  pag = pag
    .replaceAll('{{MENU}}', mp)
    .replaceAll('{{SUM}}', `<p class="sumario__titulo">${titulo}</p>\n` + topicosToc(partes0));
  pag = pag.replaceAll('{{NP}}', String(np_)).replaceAll('{{NT}}', String(nt)).replaceAll('{{PARTES}}', partes0);
  escreverTexto(arq, pag);
  console.log(`${arq}: ${np_} partes, ${nt} topicos`);
} else if (modo === 'add') {
  const arq = args[1];
  let h = lerTexto(arq);
  let partes = partes0;

  const ultimo = Math.max(...[...h.matchAll(/<section class="parte" id="s(\d+)">/g)].map((m) => Number(m[1])));
  const ultimoNum = Math.max(...[...h.matchAll(/PARTE (\d+)<\/span>/g)].map((m) => Number(m[1])));

  const secIds = [...partes.matchAll(/<section class="parte" id="s(\d+)">/g)].map((m) => m[1]);
  secIds.forEach((sec, i) => {
    partes = partes.replace(`id="s${sec}">`, `id="s@${ultimo + i + 1}">`);
  });
  const numeros = [...partes.matchAll(/PARTE (\d+)<\/span>/g)].map((m) => m[1]);
  numeros.forEach((n, i) => {
    partes = partes.replace(`PARTE ${n}</span>`, `PARTE @${ultimoNum + i + 1}</span>`);
  });
  partes = partes.replaceAll('id="s@', 'id="s').replaceAll('PARTE @', 'PARTE ');

  h = h.replaceAll('\n</main>', '\n\n' + partes + '\n</main>');
  const novoToc = topicosToc(partes);
  h = h.replaceAll('</div>\n</nav>\n<main class="leitura">', novoToc + '\n</div>\n</nav>\n<main class="leitura">');
  const np_ = (h.match(/class="parte"/g) || []).length;
  const nt = (h.match(/class="topico"/g) || []).length;
  h = h.replace(/<span>\d+ partes? · \d+ tópicos<\/span>/g, () => `<span>${np_} partes · ${nt} tópicos</span>`);
  escreverTexto(arq, h);
  console.log(`${arq}: agora ${np_} partes, ${nt} topicos`);
} else {
  console.error('uso: node scripts/mk.mjs new <arq> <titulo> <desc> <lead> <link-antes> <rotulo>   |   node scripts/mk.mjs add <arq>');
  process.exit(1);
}
